dart/lib/flat_buffers.dart

```diff
@@ -1241,3 +1246,72 @@ class _VTable {
     }
   }
 }
+
+// Allocator interface. This is flatbuffers-specific - only for [Builder].
+abstract class Allocator {
+  ByteData allocate(int size);
+
+  void deallocate(ByteData data);
+
+  /// Reallocate [newSize] bytes of memory, replacing the old [oldData]. This
+  /// grows downwards, and is intended specifically for use with [Builder].
+  /// Params [inUseBack] and [inUseFront] indicate how much of [oldData] is
+  /// actually in use at each end, and needs to be copied.
+  ByteData reallocateDownward(
+      ByteData oldData, int newSize, int inUseBack, int inUseFront) {
+    final newData = allocate(newSize);
+    clear(newData, true);
+    _copyDownward(oldData, newData, inUseBack, inUseFront);
+    deallocate(oldData);
+    return newData;
+  }
+
+  /// Clear the allocated data contents.
+  ///
+  /// Param [isFresh] is true if the given data has been freshly allocated,
+  /// depending on the allocator implementation, clearing may be unnecessary.
```

Comment:
What does fresh mean? All zeroes? Random garbage memory from a new malloc?

Maybe we should name this `isDirty` instead, or `isZeroed` or something?

Reply:
Hmm, I must try to rephrase the docs then because it clearly didn't get the point across. 

What I was trying to say was that the data was "just" allocated and haven't been used previously. This may mean, depending on the actual `Allocator` implementation that the data is already all zeroed out (e.g. with `DefaultAllocator`) or it might not be (with the sample C allocator I've linked in the description) and the allocator should do it, because Builder currently expects the buffer to be all zeroes.

Approaching this from another angle: maybe if you/someone else with FB internals knowledge could clarify whether the data actually needs to be cleared, this whole clearing stuff could go away completely, making the builder a little faster during allocation/reset().

Reply:
How is being "just" allocated meaningful though? :)

If It was "just" allocated as a Dart `ByteBuffer`, it's zeroed.

If it was "just" allocated using `calloc`, then it's zeroed.

If it was "just" allocated using `malloc`, it may or may not be zeroed (and may or may not be physically the same block of memory that was recently `free`d holding some flatbuffer data).

I don't know that this will have any special FB specific domain meaning - but it would be good for callers and implementers to have a clear idea of what this parameter means and what assumptions they can make about it. The more clear we can be about that the better.

Reply:
As you've pointed out, it is meaningful for the concrete implementation - whatever allocation method they've chosen, they'll know (or rather they can know) whether the buffer is zeroed out or needs "clearing" when the `clear()` is called.

Anyway, I see this is causing confusion already so I agree about changing it somehow... let's finish the discussion about the custom buffer class first I guess